File: app/src/main/java/xyz/rodit/snapmod/features/info/AdditionalFriendInfo.java
package xyz.rodit.snapmod.features.info;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;

import de.robv.android.xposed.XC_MethodHook;
import xyz.rodit.snapmod.Shared;
import xyz.rodit.snapmod.features.Feature;
import xyz.rodit.snapmod.features.FeatureContext;
import xyz.rodit.snapmod.mappings.CalendarDate;
import xyz.rodit.snapmod.mappings.FooterInfoItem;
import xyz.rodit.snapmod.mappings.FriendProfilePageData;
import xyz.rodit.snapmod.mappings.FriendProfileTransformer;

public class AdditionalFriendInfo extends Feature {

    public AdditionalFriendInfo(FeatureContext context) {
        super(context);
    }

    @Override
    protected void performHooks() {
        // Show more info in friend profile footer
        FriendProfileTransformer.apply.hook(new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) {
                if (context.config.getBoolean("more_profile_info") && FriendProfilePageData.isInstance(param.args[0]) && param.getResult() instanceof List) {
                    List viewModelList = (List) param.getResult();
                    if (viewModelList.isEmpty()) {
                        return;
                    }

                    Object viewModel = viewModelList.get(0);
                    if (FooterInfoItem.isInstance(viewModel)) {
                        FriendProfilePageData data = FriendProfilePageData.wrap(param.args[0]);
                        Date friendDate = new Date(Math.max(data.getAddedTimestamp(), data.getReverseAddedTimestamp()));
                        CalendarDate birthday = data.getBirthday();
                        if (birthday.isNull()) {
                            birthday = new CalendarDate(13, -1);
                        }

                        String info = "Friends with " +
                                data.getDisplayName() +
                                " since " +
                                SimpleDateFormat.getDateInstance().format(friendDate) +
                                ".\n" +
                                data.getDisplayName() +
                                "'s birthday is " +
                                birthday.getDay() +
                                " " +
                                Shared.MONTHS[birthday.getMonth() - 1] +
                                "\nFriendship: " +
                                data.getFriendLinkType().instance +
                                "\nAdded by: " +
                                data.getAddSourceTypeForNonFriend().instance;

                        FooterInfoItem.wrap(viewModel).setText(info);
                    }
                }
            }
        });
    }
}

File: app/src/main/java/xyz/rodit/snapmod/features/info/AdditionalFriendInfo.kt
package xyz.rodit.snapmod.features.info

import xyz.rodit.snapmod.Shared
import xyz.rodit.snapmod.features.Feature
import xyz.rodit.snapmod.features.FeatureContext
import xyz.rodit.snapmod.mappings.CalendarDate
import xyz.rodit.snapmod.mappings.FooterInfoItem
import xyz.rodit.snapmod.mappings.FriendProfilePageData
import xyz.rodit.snapmod.mappings.FriendProfileTransformer
import xyz.rodit.snapmod.util.after
import java.text.SimpleDateFormat
import java.util.*
import kotlin.math.max

class AdditionalFriendInfo(context: FeatureContext) : Feature(context) {

    override fun performHooks() {
        // Show more info in friend profile footer.
        FriendProfileTransformer.apply.after(context, "more_profile_info") {
            val transformer = FriendProfileTransformer.wrap(it.thisObject)
            if (!FriendProfilePageData.isInstance(transformer.data) || it.result !is List<*>) return@after

            val viewModelList = it.result as List<*>
            if (viewModelList.isEmpty()) return@after

            val viewModel = viewModelList[0]!!
            if (!FooterInfoItem.isInstance(viewModel)) return@after

            val data = FriendProfilePageData.wrap(transformer.data)
            val friendDate = Date(max(data.addedTimestamp, data.reverseAddedTimestamp))
            val birthday = if (data.birthday.isNull) CalendarDate(13, -1) else data.birthday

            val info = """Friends with ${data.displayName} since ${
                SimpleDateFormat.getDateInstance().format(friendDate)
            }.
${data.displayName}'s birthday is ${birthday.day} ${Shared.MONTHS[birthday.month - 1]}
Friendship: ${data.friendLinkType.instance}
Added by: ${data.addSourceTypeForNonFriend.instance}"""
            FooterInfoItem.wrap(viewModel).text = info
        }
    }
}